More advanced and thorough examples in docs

bmcfee opened the pull request:
This PR implements #41 #40 -- adding examples for handling NPZ files and parallelism.

bmcfee commented:
Probably a good idea to include a fully worked example in the docs, eh?  Eg data augmentation on iris via sklearn.

bmcfee commented:
Tagging @justinsalamon

cjacoby commented on docs/example2.rst:
```diff
@@ -48,28 +48,49 @@ As a concrete example, we can simulate a mixture of noisy streams with differing
-                   for sigma in [0.5, 1.0, 2.0, 4.0]]
+                   for sigma in [0, 0.5, 1.0, 2.0, 4.0]]
 
         # Build a mux stream, keeping only 2 streams alive at once
```
Here it says 2 streams, below 3. I assume this is not what you mean?

cjacoby commented on docs/example2.rst:
```diff
@@ -48,28 +48,49 @@ As a concrete example, we can simulate a mixture of noisy streams with differing
-                         k=2,    # Keep 2 streams alive at once
-                         lam=16) # Use a poisson rate of 16
+                         k=3,    # Keep 3 streams alive at once
+                         lam=64) # Use a poisson rate of 64
```
In the below text, you say "...`lam=16`, but it's now 64. Plzfix.

bmcfee replied:
Done.

justinsalamon commented:
@bmcfee happy to pitch in once im back, if needed.

ejhumphrey commented on docs/example3.rst:
```diff
@@ -0,0 +1,57 @@
+            while True:
+                # Compute the index offset
+                idx = np.random.randint(n_total - n)
+                yield dict(X=data['X'][idx:idx+n],
```
IIRC, this kind of slicing will (should?) cause memory leaks because the dictionary inherits the identifier of `data`, despite using the context manager. I believe [this demonstrates the behavior.](https://github.com/ejhumphrey/pescador-wtf/blob/master/test_memleak.py#L104)

I think the best practice here is to force the data copy before handing an ndarray to the dictionary, like `dict(X=np.array(data['X'][idx:idx + n]...`.

note: it makes me *wicked* uncomfortable how easy this is to overlook, but I don't have any answers I like. There are three that jump out at me:

- `Streamer` objects have a keyword argument like `asarray=True` which, by default, forces values in dictionaries to be reallocated as new arrays. I _think_ this would do the right thing, but I'm maybe 65% sure of that. Major win, the default behavior is sane; potential drawback, how detrimental to performance is it to blindly reallocate all matrix data?
- We provide an `@asarray` decorator, that one can apply to generators to enforce the data it yields is properly reallocated, allowing Python to fully clean up after a stream when it is deactivated. Major disadvantage here is that this should be the default behavior, but decorators are explicit add-ons, and like to be forgotten.
- We provide a `Sample` object, as a subclass of `dict` that does the right thing with the data it is handed. Again, not exactly default behavior, and is starting to feel slightly over-engineered.

any other ideas?

nit: there should be whitespace around the `+`s

bmcfee replied:
I've thought a bit more about this, and I think it's best that we not over-complicate things.  It's already complicated enough to use, and I don't like the idea of doing reallocation behind the scenes.

FWIW, this thread is helpful: https://mail.scipy.org/pipermail/numpy-discussion/2015-December/074376.html

TLDR: `npy` files support [mmap mode](https://docs.scipy.org/doc/numpy/reference/generated/numpy.load.html) while `npz` files do not.  If you care about memory leaking here, don't use `npz` files.  If you want memory mapping with structured data, use `pytables` instead.

I think the sane thing to do here is to drop this example, and note the memory mapping problem in the docs.

bmcfee commented:
> IIRC, this kind of slicing will (should?) cause memory leaks because the dictionary inherits the identifier of data, despite using the context manager. I believe this demonstrates the behavior.

I'm not sure that's a leak exactly - the reference is still active, so the GC won't (and shouldn't) free it.

However, once the context manager exits AND the last reference to the array is lost, then the GC should collect it.  If it's not, that's a bug in python.

It's worth mentioning that a buffered streamer does force the copy, so that could be an extra sanity check.

(In case it's not obvious, I really don't like doing explicit copies just to avoid apparent GC weirdness.)

bmcfee commented:
> I have also thought about making a sample object (I've done this in some of my own work), so that additionally the Sample object could use __slots__ = ['X', 'Y'], and thereby save a lot of memory.

The downside of this is that you'd be stuck with only the two pre-determined variable names. I sometimes have streamers that generate many more variables (eg for multitask learning for different Y's, or cross-modal learning for different X's).

Would it be possible to do something that works like a named tuple (dynamically set, but fixed fields), that you could still use in the way you describe above?

bmcfee commented:
I updated the npz/npy example again to discuss memory mapping issues, and while I was at it, fixed the build on RTD.  @ejhumphrey care to take another pass on this?

ejhumphrey commented:
looks great 🎉
